⚠️ Support Azure client request to other Azure Clouds

nader-ziada opened the pull request:
**What this PR does / why we need it**:
- if AZURE_ENVIRONMENT is set, will get the matching settings
- all clients use baseURI from settings

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #469 

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Support to connect to all regions where Azure is available, not just the default public Azure.
```

nader-ziada commented on cloud/scope/clients.go:
```diff
@@ -40,6 +42,8 @@ func (c *AzureClients) setCredentials(subscriptionID string) error {
 		return err
 	}
+	c.ResourceManagerEndpoint = settings.Environment.ResourceManagerEndpoint
+	c.ResourceManagerVMDNSSuffix = settings.Environment.ResourceManagerVMDNSSuffix
```
@devigned I'm not sure about this, in the sdk `GetSettingsFromFile` it doesn't return `ResourceManagerVMDNSSuffix` but its used by the cluster reconciler and it has different values based on the cloud used, is there somewhere else to get it?

nader-ziada commented:
@devigned @CecileRobertMichon @jseely  how would you all feel about just supporting the known environment names as a start? and not read environment from an endpoint or a file?

nader-ziada commented:
/test pull-cluster-api-provider-azure-e2e

CecileRobertMichon commented on templates/cluster-template.yaml:
```diff
@@ -66,7 +66,7 @@ spec:
     - content: |
         {
-          "cloud": "AzurePublicCloud",
+          "cloud": "${AZURE_ENVIRONMENT}",
```
this makes `AZURE_ENVIRONMENT` a required variable, it needs to be added to the CAPI book quickstart

nader-ziada replied:
is there a different way of doing this so it's optional?

devigned replied:
Not right now. Checkout https://github.com/kubernetes-sigs/cluster-api/pull/3052

I hesitate to say this, but perhaps, another flavor for the time being?

nader-ziada replied:
but its not in the templates? how would the flavor do that?

devigned replied:
I'm not sure I understand the question.

What I'm suggesting is that `cluster-teamplate.yaml` stays as it is with `"AzurePublicCloud"`, so there is no new variable required when deploying the default template. Secondly, create a new flavor which includes the "${AZURE_ENVIRONMENT}"` var.

Does that make sense?

nader-ziada replied:
got it, ok makes sense, will look into what needs to change

CecileRobertMichon replied:
Going to play devil's advocate here... I really don't like the idea of having separate flavors for different variable values. I really wish we could do defaulting instead. Since that's not possible, maybe we can just document that you have to change the `cloud` value to the Azure cloud you want to use since this is part of azure.json. Also keep in mind this whole thing will be an environment variable after https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/653/files#diff-1bf07101845de6da98f1c948b82c5ce0R296.

nader-ziada replied:
it currently defaults to AzurePublicCloud if its empty, or at least it should

nader-ziada replied:
Are we ok in concept to add a new required variable?

devigned replied:
Should we just wait until #653 is merged? As @CecileRobertMichon pointed out, once that is merged, there is no required variable :).

nader-ziada replied:
that's waiting for CAPI 0.3.7 I guess

devigned replied:
The problem with adding a required var is that it will a breaking change for anyone that will run `clusterctl config cluster --infrastructure=azure`. That will break any scripts using the default flavor.

It would be friendlier to users if we were to leave `"cloud": "AzurePublicCloud"` and provide documentation on how to change the value.

When #653 lands, that documentation is still valid, and we didn't break anyone in the mean time.

nader-ziada replied:
don't we still have to add AzureEnvironment env var to the capz-controller which would have to come from a secret or similar? I think instructions to update the secret value or change the deployment configuration would not be very user friendly

CecileRobertMichon replied:
For now let's keep the new var and try to make it optional later with https://github.com/kubernetes-sigs/cluster-api/issues/3189 @nader-ziada

nader-ziada commented on config/default/credentials.yaml:
```diff
@@ -8,4 +8,5 @@ data:
-  client-secret: ${AZURE_CLIENT_SECRET_B64}
\ No newline at end of file
+  client-secret: ${AZURE_CLIENT_SECRET_B64}
+  cloud-env: ${AZURE_ENVIRONMENT_B64}
```
@CecileRobertMichon this variable here?

nader-ziada commented:
@CecileRobertMichon @devigned  I think the PR is now ready for another look, 
 - changed the `AZURE_ENVIRONMENT` to be used in a ConfigMap so we don't have to encode it in the secret and this way we don't need another `_B64` variable 
 - updated the docs where I thought made sense 
 - added a PR  (on hold) for [the cluster-api book](https://github.com/kubernetes-sigs/cluster-api/pull/3212)

CecileRobertMichon commented on cloud/scope/clients_test.go:
```diff
@@ -0,0 +1,64 @@
+	. "github.com/onsi/gomega"
+)
+
+func TestGettingEnvironment(t *testing.T) {
```
what happens when `AZURE_ENVIRONMENT` is an invalid cloud name?
 eg. If I set `AZURE_ENVIRONMENT` to `AzureUAGovernmentCloud`  I should get an error rather than CAPZ trying to use PublicCloud as default and failing later because my sub isn't valid in public cloud.

nader-ziada replied:
added a test for an invalid value, good catch

CecileRobertMichon commented on cloud/scope/clients_test.go:
```diff
@@ -0,0 +1,83 @@
+			expectedDNSSuffix: "cloudapp.microsoftazure.de",
+			expectedError:     false,
+		}, "AZURE_ENVIRONMENT has an invalid value": {
+			azureEnv:             "AzureInSpace",
```
😆

nader-ziada commented:
any concerns/comments still outstanding?

devigned commented on docs/getting-started.md:
```diff
@@ -74,6 +74,18 @@ An Azure Service Principal is needed for populating the controller manifests. Th
+
+  5. Set the name of the AzureCloud to be used, the default value that would be used by most users is "AzurePublicCloud", other values are:
+
+  - ChinaCloud: "AzureChinaCloud"
```
nit: looks like an extra leading space on the list item

nader-ziada replied:
I had to do that for some reason to make the rendered page look aligned

devigned replied:
Indeed, it looks right when rendered.
